42.md

```diff
@@ -56,11 +56,23 @@ Relays MUST exclude `kind: 22242` events from being broadcasted to any client.
 
 ## Protocol flow
 
-At any moment the relay may send an `AUTH` message to the client containing a challenge. After receiving that the client may decide to
-authenticate itself or not. The challenge is expected to be valid for the duration of the connection or until a next challenge is sent by
-the relay.
+At specific moments the relay may send an `AUTH` message to the client containing a challenge, in one of the following forms:
 
-The client may send an auth message right before performing an action for which it knows authentication will be required -- for example, right
+- `["AUTH", <challenge-string>]`: No third array element. It means relay requires authentication for any resource access.
+If applicable, it MUST be sent right after WebSocket `connection`.
+Client may have already sent a message before receiving this one, in that case relay MUST also send an `AUTH` message in one of the other forms below;
+- `["AUTH", <challenge-string>, "sub:<subscription_id>"]`: It references a previous client `REQ` message with a specific subscription id as third array element.
+It means client subscription requested atleast one event that requires prior authentication to be accessed.
+MUST be sent before the `EOSE` message with same subscription id. The relay SHOULD send no events before `EOSE` because the client is expected to retry the `REQ` after authentication;
+- `["AUTH", <challenge-string>, "evt:<event_id>"]`: It references a previous client `EVENT` message with a specific event id as third array element.
+It means relay requires authentication before publishing any or atleast this kind of event.
+MUST be sent before `OK` message with same event id.
```

Comment:
the behavior here should be consistent with `REQ` auth flow. can we add "Client is expected to retry the `EVENT` after authentication"?

also maybe specify that `OK` message should have value of false.